lib/src/barback/dartdevc/dartdevc.dart

```diff
@@ -0,0 +1,231 @@
+    Module module,
+    ScratchSpace scratchSpace,
+    Set<AssetId> linkedSummaryIds,
+    Map<String, String> environmentConstants,
+    BarbackMode mode,
+    logError(String message)) {
+  var outputCompleters = <AssetId, Completer<Asset>>{
+    module.id.jsId: new Completer<Asset>(),
+  };
+  if (mode == BarbackMode.DEBUG) {
+    outputCompleters[module.id.jsSourceMapId] = new Completer<Asset>();
+  }
+
+  () async {
+    var jsOutputFile = scratchSpace.fileFor(module.id.jsId);
+    var sdk_summary = p.url.join(sdkDir.path, 'lib/_internal/ddc_sdk.sum');
+    var request = new WorkRequest();
+    request.arguments.addAll([
+      '--dart-sdk-summary=$sdk_summary',
+      // TODO(jakemac53): Remove when no longer needed,
+      // https://github.com/dart-lang/pub/issues/1583.
+      '--unsafe-angular2-whitelist',
+      '--modules=amd',
+      '--dart-sdk=${sdkDir.path}',
+      '--module-root=${scratchSpace.tempDir.path}',
+      '--library-root=${p.dirname(jsOutputFile.path)}',
+      '--summary-extension=${linkedSummaryExtension.substring(1)}',
+      '--no-summarize',
+      '-o',
+      jsOutputFile.path,
+    ]);
+
+    if (mode == BarbackMode.RELEASE) {
+      request.arguments.add('--no-source-map');
+    }
+
+    // Add environment constants.
+    environmentConstants.forEach((key, value) {
+      request.arguments.add('-D$key=$value');
+    });
+
+    // Add all the linked summaries as summary inputs.
+    for (var id in linkedSummaryIds) {
+      request.arguments.addAll(['-s', scratchSpace.fileFor(id).path]);
+    }
+
+    // Add URL mappings for all the package: files to tell DartDevc where to find
+    // them.
+    for (var id in module.assetIds) {
+      var uri = canonicalUriFor(id);
+      if (uri.startsWith('package:')) {
+        request.arguments
+            .add('--url-mapping=$uri,${scratchSpace.fileFor(id).path}');
+      }
+    }
+    // And finally add all the urls to compile, using the package: path for files
+    // under lib and the full absolute path for other files.
+    request.arguments.addAll(module.assetIds.map((id) {
+      var uri = canonicalUriFor(id);
+      if (uri.startsWith('package:')) {
+        return uri;
+      }
+      return scratchSpace.fileFor(id).path;
+    }));
+
+    var response = await dartdevcDriver.doWork(request);
+
+    // TODO(jakemac53): Fix the ddc worker mode so it always sends back a bad
+    // status code if something failed. Today we just make sure there is an output
+    // js file to verify it was successful.
+    if (response.exitCode != EXIT_CODE_OK || !jsOutputFile.existsSync()) {
+      var message =
+          'Error compiling dartdevc module: ${module.id}.\n${response.output}';
+      logError(message);
+      outputCompleters.values.forEach((completer) {
+        completer.completeError(message);
+      });
+    } else {
+      outputCompleters[module.id.jsId].complete(
+          new Asset.fromBytes(module.id.jsId, jsOutputFile.readAsBytesSync()));
```

Comment:
Can you just return `Asset.fromPath()` here and delay reading the file off disc until needed?